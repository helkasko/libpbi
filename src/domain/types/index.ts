import * as pbi from 'powerbi-client';

export enum ReportType {
  VISUAL = 'visual',
  REPORT = 'report',
}

export type AbstractReportConfig = ReportConfig & {
  type: ReportType;
  tokenType: pbi.models.TokenType;
};

export type ReportConfig = {
  description?: string;
  reportId: string;
  groupId: string;
};

export type EmbedTokenReportConfig = ReportConfig & {
  accessToken: string;
};

export type EmbedTokenVisualConfig = EmbedTokenReportConfig & {
  pageName: string;
  visualName: string;
};
