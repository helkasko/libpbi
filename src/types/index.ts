import * as pbi from 'powerbi-client';

export enum ReportType {
  VISUAL = 'visual',
  REPORT = 'report',
}

export type EmbedConfig = {
  [key: string]: string | ReportType | pbi.models.TokenType | undefined;
  type: ReportType;
  tokenType: pbi.models.TokenType;
  embedUrl?: string;
};

export type ReportConfig = {
  description?: string;
  reportId: string;
  groupId: string;
};

export type EmbedTokenReportConfig = ReportConfig & {
  accessToken: string;
};

export type EmbedTokenVisualConfig = EmbedTokenReportConfig & {
  pageName: string;
  visualName: string;
};
