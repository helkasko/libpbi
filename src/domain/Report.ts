import { AbstractReportConfig } from './types';

/**
 * The root Report class. Designed to be extended by more specific report classes.
 */
export default class Report {
  private readonly config: AbstractReportConfig;

  constructor(config: AbstractReportConfig) {
    this.config = {
      ...config,
      embedUrl: `https://app.powerbi.com/reportEmbed?reportId=${config.id}&groupId=${config.groupId}`,
    };
  }

  /**
   * Returns the configuration of a report to be passed to Power BI interface when embedding a report
   */
  getConfig(): AbstractReportConfig {
    return this.config;
  }
}
