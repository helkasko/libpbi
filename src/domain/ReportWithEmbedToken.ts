import * as pbi from 'powerbi-client';
import Report from './Report';
import { EmbedTokenReportConfig, ReportType } from './types';

/**
 * Represents a Power BI report that is configured with an embed token.
 */
export default class ReportWithEmbedToken extends Report {
  constructor(config: EmbedTokenReportConfig) {
    super({
      ...config,
      tokenType: pbi.models.TokenType.Embed,
      type: ReportType.REPORT
    });
  }
}
