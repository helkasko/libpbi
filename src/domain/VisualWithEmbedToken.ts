import * as pbi from 'powerbi-client';
import Report from './Report';
import { EmbedTokenVisualConfig, ReportType } from '../types';

/**
 * Represents a Power BI visual that is configured with an embed token
 */
export default class VisualWithEmbedToken extends Report {
  constructor(config: EmbedTokenVisualConfig) {
    super({
      tokenType: pbi.models.TokenType.Embed,
      type: ReportType.VISUAL
    }, config);
  }
}
