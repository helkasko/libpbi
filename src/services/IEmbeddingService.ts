import * as pbi from 'powerbi-client';
import { AbstractReportConfig } from '../domain/types';

/**
 * Defines an interface for an embedding service for Power BI reports
 */
export default abstract class IEmbeddingService {
  abstract embed: (element: HTMLElement, config: AbstractReportConfig) => pbi.Embed | null;
}
