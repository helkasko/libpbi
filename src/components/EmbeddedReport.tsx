import IEmbeddingService from '../services/IEmbeddingService';
import PowerBIEmbeddingService from '../services/PowerBIEmbeddingService';
import * as React from 'react';
import Report from '../domain/Report';

type ReportProps = {
  report: Report;
  embeddingService?: IEmbeddingService;
};

/**
 * Renders an embedded PowerBI report.
 */
export default class EmbeddedReport extends React.PureComponent<ReportProps> {
  reportRef: React.RefObject<HTMLDivElement>;
  embeddingService: IEmbeddingService;

  constructor(props: ReportProps) {
    super(props);

    this.reportRef = React.createRef();
  }

  componentDidMount() {
    this.getEmbeddedReport();
  }

  private getEmbeddedReport(): void {
    if (!this.reportRef.current) {
      return;
    }

    const { report } = this.props;
    const embeddingService = this.props.embeddingService || PowerBIEmbeddingService;
    embeddingService.embed(this.reportRef.current, report.getConfig());
  }

  render() {
    return <div ref={this.reportRef} />;
  }
}
