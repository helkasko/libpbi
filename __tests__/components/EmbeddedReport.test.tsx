import EmbeddedReport from '../../src/components/EmbeddedReport';
import { shallow } from 'enzyme';
import IEmbeddingService from '../../src/services/IEmbeddingService';
import * as React from 'react';
import ReportWithEmbedToken from '../../src/domain/ReportWithEmbedToken';
import { AbstractReportConfig, EmbedTokenReportConfig } from '../../src/domain/types';

const dummyEmbeddingService: IEmbeddingService = {
  embed: (_: HTMLElement, __: AbstractReportConfig): null => {
    return null;
  }
};

const config: EmbedTokenReportConfig = {
  accessToken: 'dumdum',
  description: 'lollol',
  groupId: 'groupId',
  reportId: 'reportId'
};

const report = new ReportWithEmbedToken(config);

describe('<EmbeddedReport />', () => {
  it('renders correctly', () => {
    const component = shallow(<EmbeddedReport report={report} embeddingService={dummyEmbeddingService} />);
    expect(component).toMatchSnapshot();
  });
});
