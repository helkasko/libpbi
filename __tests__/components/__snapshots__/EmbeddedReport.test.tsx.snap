// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmbeddedReport /> renders correctly 1`] = `<div />`;
